x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/ApiKeyService.java

```diff
@@ -433,23 +431,31 @@ static XContentBuilder newDocument(
             .field("api_key_invalidated", false);
 
         addApiKeyHash(builder, apiKeyHashChars);
-        addRoleDescriptors(builder, keyRoles);
-        addLimitedByRoleDescriptors(builder, userRoles);
+        addRoleDescriptors(builder, keyRoleDescriptors);
+        addLimitedByRoleDescriptors(builder, userRoleDescriptors);
 
         builder.field("name", name).field("version", version.id).field("metadata_flattened", metadata);
         addCreator(builder, authentication);
 
         return builder.endObject();
     }
 
-    static XContentBuilder buildUpdatedDocument(
+    // package private for testing
+
+    /**
+     * @return `null` if the update is a noop, i.e., if no changes to `currentApiKeyDoc` are required
+     */
+    XContentBuilder buildUpdatedDocument(
```

Comment:
As suggested in the other comment: I think it's helpful to rename this to `maybeBuildUpdatedDocument`.